### PR TITLE
Remove mage notice in favour of make notice

### PR DESCRIPTION
**Pull request by belimawr**
## What this PR actually do?
It removes the `mage notice` target in favour of the `make notice` rule. Please read the PR's body and all comments for a full history/context/reasoning.

## What does this PR do?
Fix a race condition when calling `go-licence-detector`. The original code waited for a goroutine to write to `go-licence-detector`'s stdin before starting the command, so the mage target was hanging.

This commit fixes it, adds logging (when verbose set) to the command run as well as improves the error message.

## Why is it important?
`mage notice` is used to update licences, which is required whenever we update or add dependencies.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- [ ] ~~I have made corresponding change to the default configuration files~~
- [ ] ~~I have added tests that prove my fix is effective or that my feature works~~
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist
- [x] There is no `CHANGELOG-developer.next.asciidoc`, should this fix be mentioned on `CHANGELOG.next.asciidoc`?

## How to test this PR locally
Run `mage notice` from the repo root, it won't hang and if there is an error, it will print it out.

## Related issues

- Blocks #689 
- Relates #1107 

~~## Use cases~~
~~## Screenshots~~
~~## Logs~~

### CODE REVIEW

**Review by ChrsMark — APPROVED** [review]
lgtm

File: CHANGELOG.next.asciidoc [hunk 1]
```diff
@@ -20,6 +20,7 @@
 - Remove username/password for fleet-server authentication. {pull-beats}[29458]
 
 ==== Bugfixes
+- Fix race condition on `mage notice` {pull}[1108]
```
Comment (cmacknz) [inline CHANGELOG.next.asciidoc:23]:
```suggestion
```

You don't need a changelog entry for this as it isn't user facing.

**Comment by cmacknz** [comment]
Could we just replace this implementation with the one in elastic-agent-libs? https://github.com/elastic/elastic-agent-libs/blob/ee6ce69b6f17a29307878527768f687d4d77faef/magefile.go#L83

**Comment by belimawr** [comment]
> Could we just replace this implementation with the one in elastic-agent-libs? https://github.com/elastic/elastic-agent-libs/blob/ee6ce69b6f17a29307878527768f687d4d77faef/magefile.go#L83

I'll give it a shot today.

**Comment by belimawr** [comment]
Removing the last argument `-depsOut ""` seems to have done the trick. Maybe the "empty string" wasn't properly escaped and it was confusing `go-licence-detector`.

Btw, I tried to use the implementation on `elstic-agent-libs`, but it kept complaining about missing overrides for licences. I added one or two manually but it kept complaining, so I gave up and tried the original implementation again.

**Review by michalpristas — APPROVED** [review]

**Comment by belimawr** [comment]
Thanks to our CI, I've discovered that we have **two different implementations** to generate the NOTICE.txt file: one in the `magefile.go` and another in the `Makefile`. The Makefile one adds something at the end and it's the one used on CI.

I'll make them behave the same for now, but we should look into removing this Makefile logic.

**Comment by narph** [comment]
@belimawr , the magefile  `Notice` func was an initial attempt to replace the make notice function, I should have added in the comment that the function will not work and we should re-use the function  from `elastic-agent-libs` https://github.com/elastic/elastic-agent-libs/blob/ee6ce69b6f17a29307878527768f687d4d77faef/magefile.go#L83 as we are doing for the shipper.
I recommend re-using that instead of adding a different solution that we need to maintain.

**Comment by cmacknz** [comment]
The license overrides are normal. There is an existing overrides file for the agent but perhaps the path just needs to change or be updated as it is different between the shipper and agent:
- https://github.com/elastic/elastic-agent-shipper/blob/main/dev-tools/templates/notice/overrides.json
- https://github.com/elastic/elastic-agent-shipper/blob/main/dev-tools/templates/notice/overrides.json

**Comment by cmacknz** [comment]
I also just noticed that CI builds use `make check-ci` which uses `make notice` instead of `mage notice`. We should have a single way to do things, probably standardizing on mage as we do everywhere else.

**Comment by belimawr** [comment]
Folks, thanks a lot for all the info!

I've tried once more to get the mage notice from `elastic-agent-libs` working, but I can't seem to figure out why it won't work.

I pushed one more commit that contains the "fixed" mage notice as well as one from elastic-agent-libs (called `noticelib`).

Here is what happens when I run them and `make notice`:
```
tiago@millennium-falcon elastic-agent  v1.19  fix-magefile-notice [$!?] % mage -v noticelib
Running target: NoticeLib
Running dependency: github.com/elastic/elastic-agent-libs/dev-tools/mage.Deps.CheckModuleTidy
Running dependency: github.com/elastic/elastic-agent-libs/dev-tools/mage.InstallGoNoticeGen
exec: go "mod" "tidy"
exec: go "install" "go.elastic.co/go-licence-detector"
exec: git "diff" "--exit-code" "go.mod"
exec: go "mod" "download" "all"
exec: go "list" "-json" "-m" "all"
exec: go-licence-detector "-in" "/tmp/depsout863927292" "-includeIndirect" "-overrides" "dev-tools/notice/overrides.json" "-rules" "dev-tools/notice/rules.json" "-noticeTemplate" "dev-tools/notice/NOTICE.txt.tmpl" "-noticeOut" "NOTICE.txt"
2022/09/08 19:59:18 Failed to detect licences: no licence file found for github.com/acomagu/bufpipe. Add an override entry with licence type to continue.
Error: running "go-licence-detector -in /tmp/depsout863927292 -includeIndirect -overrides dev-tools/notice/overrides.json -rules dev-tools/notice/rules.json -noticeTemplate dev-tools/notice/NOTICE.txt.tmpl -noticeOut NOTICE.txt" failed with exit code 1
tiago@millennium-falcon elastic-agent  v1.19  fix-magefile-notice [$!?] % mage -v notice   
Running target: Notice
>> Generating NOTICE
>> fmt - go mod tidy
exec: go "mod" "tidy" "-v"
>> fmt - go mod download
exec: go "mod" "download"
>> fmt - go list
exec: go "list" "-m" "-json" "all"
>> fmt - go run
exec: go "run" "go.elastic.co/go-licence-detector" "-includeIndirect" "-rules" "dev-tools/notice/rules.json" "-overrides" "dev-tools/notice/overrides.json" "-noticeTemplate" "dev-tools/notice/NOTICE.txt.tmpl" "-noticeOut" "NOTICE.txt"
tiago@millennium-falcon elastic-agent  v1.19  fix-magefile-notice [$?] % make notice
Generating NOTICE
go mod tidy
go mod download
go list -m -json all | go run go.elastic.co/go-licence-detector \
        -includeIndirect \
        -rules dev-tools/notice/rules.json \
        -overrides dev-tools/notice/overrides.json \
        -noticeTemplate dev-tools/notice/NOTICE.txt.tmpl \
        -noticeOut NOTICE.txt \
        -depsOut ""
cat dev-tools/notice/NOTICE.txt.append >> NOTICE.txt
tiago@millennium-falcon elastic-agent  v1.19  fix-magefile-notice [$?] % 
```

Aside the fact that `elastic-agent-libs` pipes the output of `go list` (which should not affect the output), the rest seems to be the same... I even grabbed the temporary files and output piped by `elastic-agent-libs`, but I can't seem to find a reason for this different result.

**Comment by belimawr** [comment]
To avoid spending much more time on this issue (as we have a working implementation), I talked with @cmacknz and decided on keeping only the make implementation for the time being.

I also opened an issue to migrate it to mage and implement all necessary fixes: https://github.com/elastic/elastic-agent/issues/1141.

This PR now removes the `mage notice` target and documents the usage of `make notice`.

**Review by cmacknz — APPROVED** [review]

**Comment by ChrsMark** [comment]
Can we have this merged please? FF is getting closer and other PRs are blocked on this.

**Comment by belimawr** [comment]
> Can we have this merged please? FF is getting closer and other PRs are blocked on this.

Sorry the delay, CI was flaky, to run and now the build seems to be broken on MacOS. I'm looking into this right now.

**Comment by belimawr** [comment]
It seems we're having some issues with the MacOS workers, I'll merge this PR anyways because it's doc + removing something that was already not working.

If something else breaks down the line, we can revert this PR.

**Comment by cmacknz** [comment]
The MacOS failures are a known issue unfortunately, they can be ignored.

**Pull request merged**